app/forms/declaration/additionaldocuments/DocumentWriteOff.scala

```diff
@@ -16,44 +16,79 @@
   val measurementUnitKey = "measurementUnit"
+  val qualifierKey = "qualifier"
   val documentQuantityKey = "documentQuantity"
 
+  private def form2Model: (Option[String], Option[String], Option[String]) => DocumentWriteOff = {
+    case (None, None, quantity)      => convert(None, quantity)
+    case (unit, None, quantity)      => convert(unit, quantity)
+    case (unit, qualifier, quantity) => convert(Some(Seq(unit, qualifier).flatten.mkString("#")), quantity)
+  }
+
+  private def model2Form: DocumentWriteOff => Option[(Option[String], Option[String], Option[String])] =
+    model => {
+      val unitAndQualifier: Option[Array[String]] = model.measurementUnit.map(_.split("#"))
+      Some(unitAndQualifier.flatMap(_.headOption), unitAndQualifier.flatMap(_.lift(1)), model.documentQuantity.map(_.toString()))
+    }
+
   val mapping = Forms
     .mapping(
       measurementUnitKey -> optional(
         text()
-          .verifying("supplementary.addDocument.measurementUnit.error.length", lengthInRange(measurementUnitMinLength)(measurementUnitMaxLength))
-          .verifying("supplementary.addDocument.measurementUnit.error.specialCharacters", isAlphanumericWithSpecialCharacters(Set('#')))
+          .verifying("declaration.addDocument.measurementUnit.error", hasSpecificLength(measurementUnitLength) and isAlphanumeric)
+      ),
+      qualifierKey -> optional(
+        text()
+          .verifying("declaration.addDocument.qualifier.error", hasSpecificLength(qualifierLength) and isAlphanumeric)
       ),
       documentQuantityKey -> optional(
-        bigDecimal
-          .verifying("supplementary.addDocument.documentQuantity.error.precision", _.precision <= documentQuantityMaxLength)
-          .verifying("supplementary.addDocument.documentQuantity.error.scale", _.scale <= documentQuantityMaxDecimalPlaces)
-          .verifying("supplementary.addDocument.documentQuantity.error", _ >= 0)
+        text()
+          .verifying(
+            "declaration.addDocument.documentQuantity.error",
+            input => input.isEmpty || noLongerThan(documentQuantityMaxLength)(input.replaceAll("\\.", ""))
+          )
+          .verifying(
+            "declaration.addDocument.documentQuantity.error",
+            isEmpty or isDecimalWithNoMoreDecimalPlacesThan(documentQuantityMaxDecimalPlaces)
+          )
       )
-    )(DocumentWriteOff.apply)(DocumentWriteOff.unapply)
-    .verifying("supplementary.addDocument.error.measurementUnitAndQuantity", validateMeasurementUnitAndDocumentQuantity(_))
-
-  private def validateMeasurementUnitAndDocumentQuantity(documentWriteOff: DocumentWriteOff): Boolean =
-    (documentWriteOff.measurementUnit.isEmpty && documentWriteOff.documentQuantity.isEmpty) ||
-      (documentWriteOff.measurementUnit.nonEmpty && documentWriteOff.documentQuantity.nonEmpty)
+    )(form2Model)(model2Form)
 
   def form(): Form[DocumentWriteOff] = Form(mapping)
+
+  def validate(writeOff: DocumentWriteOff): Seq[FormError] = {
```

Comment:
I think this method can have something with error in name, I mean `def globalErrors` or `def additionalErrors` name validate is saying that you're just validating, but this methos is creating a new errors